File: backend/routers/aefi.py
"""Endpoints AEFI (eventos adversos post-vacunacion)."""
from typing import Optional
from fastapi import APIRouter, Depends, Query, Request
from pydantic import BaseModel, Field
from core.database import get_db
from core.constants import AEFI_VENTANA_DIAS
from services import aefi_service, audit_service

router = APIRouter(prefix="/aefi", tags=["aefi"])


class AefiIn(BaseModel):
    dosis_id: int
    sintomas: str = Field(..., min_length=3)
    severidad: str = "leve"
    inicio_minutos: Optional[int] = None
    requiere_seguimiento: bool = False


@router.get(
    "/usuario/{curp}",
    summary="Reportes AEFI del ciudadano",
)
def por_usuario(curp: str, db = Depends(get_db)):
    return aefi_service.listar_por_usuario(db, curp)


@router.get(
    "/usuario/{curp}/dosis-reportables",
    summary="Dosis elegibles para reportar AEFI",
    description=(
        f"Solo se pueden reportar reacciones dentro de los {AEFI_VENTANA_DIAS} "
        "dias posteriores a la aplicacion de la dosis. Este endpoint devuelve la "
        "lista filtrada para alimentar el selector del frontend."
    ),
)
def dosis_reportables(curp: str, db = Depends(get_db)):
    """Dosis del usuario que aun estan dentro de la ventana para reportar AEFI."""
    return {
        "ventana_dias": AEFI_VENTANA_DIAS,
        "dosis": aefi_service.dosis_reportables(db, curp),
    }


@router.get(
    "",
    summary="Listar reportes AEFI (admin)",
)
def listar(
    severidad: Optional[str] = Query(None, description="Filtra por severidad: leve, moderada, severa, grave"),
    limit: int = Query(200, ge=1, le=1000),
    db = Depends(get_db),
):
    return aefi_service.listar_para_admin(db, severidad, limit)


@router.post(
    "",
    summary="Reportar evento adverso (AEFI)",
    description=(
        "Valida que la dosis se haya aplicado dentro de los "
        f"{AEFI_VENTANA_DIAS} dias previos. Severidades permitidas: "
        "`leve`, `moderada`, `severa`, `grave`."
    ),
)
def reportar(data: AefiIn, request: Request, db = Depends(get_db)):
    new_id = aefi_service.reportar(
        db,
        dosis_id=data.dosis_id,
        sintomas=data.sintomas,
        severidad=data.severidad,
        inicio_minutos=data.inicio_minutos,
        requiere_seguimiento=data.requiere_seguimiento,
    )
    audit_service.log_event(
        db, actor_curp=None, accion="reportar_aefi",
        recurso="aefi_reportes", recurso_id=new_id, request=request,
        datos={"severidad": data.severidad, "dosis_id": data.dosis_id},
    )
    return {"ok": True, "id": new_id}

File: mobile/src/api/aefi.ts
/**
 * API AEFI (eventos adversos post-vacunacion).
 *
 * Endpoints expuestos por `backend/routers/aefi.py`:
 *  - GET  /aefi/usuario/{curp}                 -> Reporte[]
 *  - GET  /aefi/usuario/{curp}/dosis-reportables -> { ventana_dias, dosis }
 *  - POST /aefi                                -> { ok, id }
 */
import { apiClient } from './client'

export type SeveridadAefi = 'leve' | 'moderada' | 'severa' | 'grave'

export type AefiReporteUsuario = {
  id: number
  dosis_id: number
  sintomas: string
  severidad: SeveridadAefi
  inicio_minutos: number | null
  requiere_seguimiento: number // 1 / 0
  vacuna_nombre: string
  fecha_aplicacion: string
  creado_en: string
}

export type DosisReportable = {
  id: number
  vacuna_nombre: string
  numero_dosis: number
  fecha_aplicacion: string
  dias_transcurridos: number
}

export type DosisReportablesPayload = {
  ventana_dias: number
  dosis: DosisReportable[]
}

export type ReportarAefiPayload = {
  dosis_id: number
  sintomas: string
  severidad: SeveridadAefi
  inicio_minutos?: number
  requiere_seguimiento: boolean
}

export const aefiApi = {
  listarPorUsuario: (curp: string) =>
    apiClient.get<AefiReporteUsuario[]>(`/aefi/usuario/${curp}`),

  dosisReportables: (curp: string) =>
    apiClient.get<DosisReportablesPayload>(
      `/aefi/usuario/${curp}/dosis-reportables`,
    ),

  reportar: (data: ReportarAefiPayload) =>
    apiClient.post<{ ok: true; id: number }>('/aefi', data),
}
